107 simulation kyivstar data

edarin opened the pull request:
First version of simulated data.

Five files are created in `out_dir, "simulation", "Kyivstar"`:
1. "kyvstar_pop_observed.csv"
2. "kyvstar_flows_observed.csv"
3. "kyvstar_flows.csv"
4. "kyvstar_pop.csv"
5. "kyvstar_transitions.csv"

Here are the steps to simulate true population:

- population by age sex at time t0 > transition probability > population at time t > flow at time t

Here are the steps to simulate Kyivstar data:

- population detection probability by age and sex > detected population > observed population (with an added normally distributed noise)
- flow detection probability by destination, age and sex > detected flows > observed flows (with an added normally distributed noise)

alaparicioc committed:
Create KyivStar_data.R

edarin committed:
first draft simulation

edarin committed:
Merge branch 'dev' into 107-simulation-kyivstar-data

edarin committed:
add flows and observations model for simulation

edarin committed:
add comment to simulation script

edarin committed:
separate by age and sex

doug-leasure commented on src/simulation/1_simulate_data/KyivStar_data.R:
```diff
@@ -0,0 +1,244 @@
+  arrange(time, age, sex, destination) |>
+  bind_cols(transition)
+
+pop_data <- transition_data |>
```
I'm finding it a little difficult to follow the matrix operations when the data are stored as long-format tibbles for dplyr. I'm sure you've got it all figured out, but just a note that it is a little difficult for me to check the matrix operations.

doug-leasure commented on src/simulation/1_simulate_data/KyivStar_data.R:
```diff
@@ -0,0 +1,244 @@
+  left_join(pop_detection_prob) |>
+  rowwise() |>
+  mutate(
+    pop_detected = pop * prop,
```
Why use normal error for a binomial or poisson process? 
Maybe:
`users <- rpois(1, pop * prop)`
or
`users <- rbinom(1, pop, prop)`
?

doug-leasure commented on src/simulation/1_simulate_data/KyivStar_data.R:
```diff
@@ -0,0 +1,244 @@
+  ) |>
+  select(-prop_values, -prop) |>
+  rowwise() |>
+  mutate(across(locations, ~ abs(rnorm(1, .x, flow_noise))))
```
A stochastic poisson or binomial observation process may be more appropriate here instead of adding the absolute value of a Gaussian error term. A lognormal would be another distribution that may be appropriate.

edarin replied:
I chose lognormal

doug-leasure commented on src/simulation/1_simulate_data/KyivStar_data.R:
```diff
@@ -0,0 +1,244 @@
+
+
+baseline_pop$pop <- round(
+  rnorm(nrow(baseline_pop), mean = 3000, sd = 50)
```
Why not `rlnorm(n, log(3000), 0.05 / 2)`? This would give the correct range of support and make the variance scale with the mean.

doug-leasure commented on src/simulation/1_simulate_data/KyivStar_data.R:
```diff
@@ -0,0 +1,272 @@
+
+# add observation error
+flow_noise <- 5
+observed_flows <- flows |>
```
It would be good to back-calculate flows from baseline (rather than t-1) as well. This seemed to be important data based on our initial discussion on modelling strategy.

doug-leasure reviewed (changes requested):
Thanks for this! The plots are really helpful to see what's going on. It looks like you put some effort into making realistic trends. I left individual comments on specific lines of code for you to consider.

The only requested change would be to generate flows data relative to t=0 in addition to the flows relative to t-1.

edarin committed:
change simulation method from tidyverse to matrix. make it not dependent on kyivstar

edarin commented:
I have integrated all changes.
I have added the baseline flows.
I have added an abroad category.
I have made mobile phone company agnostic.
I have changed simulation strategy from tibble to matrices (way smarter).
I have changed the distributions:
- Poisson for the true baseline population stocks
- Scaled normal for the migration rate (maybe i should change that?)
- Binomial for the observed population stocks, flows and baseline flows.

Here are the current patterns:
```
For the baseline population:
locations <- sort(c("Kyiv", "Lviv", "Odesa", "Kharkiv", "Abroad")) # 5 locations
ages <- c("20-29", "30-39", "40-49", "50-59", "60+") # 7 age groups
sexes <- c("Female", "Male") 
location_shares <- c(0.05, 0.20, 0.35, 0.2, 0.2)
sex_shares <- c(0.5, 0.5)
age_shares <- c(0.16, 0.19, 0.19, 0.21, 0.25)

For the true migration probability:
# - more stayers
# - more out-migration from Kharkiv
# - even more out-migration from Kharkiv for females
# - less out-migration to abroad
# - no out-migration to abroad for males
# - rate constant through time

For the detection probability of population stocks;
# varying by age and sex
# lower Abroad
# lowest in the East
# constant through time

For the detection probability of population flows
# varying by age and sex
# lower Abroad
# lowest in the East
# constant through time
# lower than stocks detection

For the detection probability of baseline population flows
# varying by age and sex
# lower than flows except for East
# constant through time
```

doug-leasure reviewed (approved):
Nice! Thanks for diving into the matrix :)
